### PR TITLE
GAWB 3026 - give clusterserviceaccount access to dataproc bucket

### PR DESCRIPTION
**Pull request by vkumra-broad**
Swat test to follow.

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala [hunk 1]
```diff
@@ -122,7 +122,8 @@ class GoogleDataprocDAO(protected val leoServiceAccountEmail: WorkbenchEmail,
     }.recover { case CallToGoogleApiFailedException(_, _, _, _) => {
       logger.error(s"Unable to authorize token: $accessToken")
       throw AuthorizationError()
-    }}
+    }
+    }
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: nit: leave this how it was

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala [hunk 2]
```diff
@@ -376,7 +377,12 @@ class GoogleDataprocDAO(protected val leoServiceAccountEmail: WorkbenchEmail,
 
   /* set the notebook service account as the staging bucket owner */
   override def setStagingBucketOwnership(cluster: LeoCluster): Future[Unit] = {
-    cluster.serviceAccountInfo.notebookServiceAccount match {
+    setBucketOwnership(cluster.serviceAccountInfo.clusterServiceAccount, cluster.googleProject, cluster.clusterName)
+    setBucketOwnership(cluster.serviceAccountInfo.notebookServiceAccount, cluster.googleProject, cluster.clusterName)
+  }
+
+  private def setBucketOwnership(serviceAccount: Option[WorkbenchEmail], googleProject: GoogleProject, clusterName: ClusterName): Future[Unit] = {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: nit: new name is a bit misleading here

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala [hunk 3]
```diff
@@ -376,7 +377,12 @@ class GoogleDataprocDAO(protected val leoServiceAccountEmail: WorkbenchEmail,
 
   /* set the notebook service account as the staging bucket owner */
   override def setStagingBucketOwnership(cluster: LeoCluster): Future[Unit] = {
-    cluster.serviceAccountInfo.notebookServiceAccount match {
+    setBucketOwnership(cluster.serviceAccountInfo.clusterServiceAccount, cluster.googleProject, cluster.clusterName)
+    setBucketOwnership(cluster.serviceAccountInfo.notebookServiceAccount, cluster.googleProject, cluster.clusterName)
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: you are dropping the first future on the floor here. you want to use something like future.traverse. also what happens if the cluster SA and notebook SA are the same? you probably want to put them into a set

tldr, something like:

```
Future.traverse ( Set( clusterServiceAccount, notebookServiceAccount ) ) { sa =>
  setTheOwnership(account, project, clusterName)
}
```

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala [hunk 4]
```diff
@@ -397,10 +403,18 @@ class GoogleDataprocDAO(protected val leoServiceAccountEmail: WorkbenchEmail,
...
+          }
+          stagingBucket <- OptionT.fromOption {
+            getStagingBucket(dCluster)
+          }
+          _ <- OptionT.liftF[Future, BucketAccessControl] {
+            executeGoogleRequestAsync(googleProject, s"Bucket ${stagingBucket.name}", bacInsert(stagingBucket))
+          }
+          _ <- OptionT.liftF[Future, ObjectAccessControl] {
+            executeGoogleRequestAsync(googleProject, s"Bucket ${stagingBucket.name}", doacInsert(stagingBucket))
+          }
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: i think the old whitespacing was more readable here

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala (hunk 3 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: +1, that'd work. Another nitpick: we could handle the Options here and make `setBucketOwnership` take a `WorkbenchEmail` instead of an Option, since the None case is a no-op. Something like:
```
Future.traverse(Set(clusterServiceAccount, notebookServiceAccount).flatten) { sa =>
  setBucketOwnership(sa, project, clusterName)
}.void
```

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala (hunk 2 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: Yeah we could keep the name `setStagingBucketOwnership`?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala (hunk 4 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: +1, these monad transformers are annoyingly verbose at times, but I usually try to keep for-comp `<-`s on 1 line if possible.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala (hunk 3 above)
Comment (helgridly) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: my inner monologue:

"wait, what does set.flatten do?"
"ohhhhh"
"goddamn."

i like this approach but it deserves a comment along the lines of "flatten any nones out of the set" for the slow-witted (i.e. me)

**Comment by helgridly** [comment]
I'll take responsibility for that - I asked for a swat test to see if we could reproduce the behavior that led to DaveS' issue. Right now I _don't_ trust Dataproc to set the right acls on the staging bucket, otherwise we wouldn't have seen this bug!

**Comment by vkumra-broad** [comment]
Created GAWB-3047 for the swat test.

**Comment by rtitle** [comment]
Reminder that dev is frozen for the AoU demo, so hold off on merging.
